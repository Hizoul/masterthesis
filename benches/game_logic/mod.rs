pub mod tetromino;
pub mod field;
pub mod transposition;