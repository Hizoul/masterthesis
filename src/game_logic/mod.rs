pub mod val;
pub mod cache;
pub mod tetromino;
pub mod log;
pub mod field;