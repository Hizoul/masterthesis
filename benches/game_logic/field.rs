extern crate criterion;

use criterion::Criterion;
use rustyblocks::game_logic::field::{GameField};
use rustyblocks::game_logic::field::helper_structs::{GameRules};
use rustyblocks::game_logic::log::{GameLog};
use rustyblocks::game_logic::cache::{ShapeCache};
use std::sync::{Arc};
use rand::prelude::*;

const BLA: &'static str = "{\"log\":[{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":0,\"x\":8,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":0,\"x\":6,\"y\":0}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":0,\"x\":4,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":0,\"x\":7,\"y\":2}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":6,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":7,\"y\":3}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":6,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":6,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":6,\"y\":14}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":1,\"x\":7,\"y\":5}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":9,\"y\":3}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":3,\"x\":8,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":10,\"y\":6}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":2,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":8,\"y\":7}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":2,\"x\":4,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":7,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":0,\"x\":4,\"y\":3}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":2,\"y\":6}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":3,\"x\":5,\"y\":5}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":3,\"x\":4,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":3,\"x\":8,\"y\":11}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":1,\"x\":4,\"y\":9}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":3,\"x\":7,\"y\":12}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":1,\"x\":0,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":4,\"y\":12}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":9,\"y\":12}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":14}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":3,\"x\":0,\"y\":3}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":0,\"y\":4}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":1,\"x\":4,\"y\":14}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":7,\"x\":0,\"y\":5}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":7,\"x\":1,\"y\":8}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":16}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":4}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":3,\"x\":5,\"y\":15}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":8}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":18}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":1,\"x\":2,\"y\":12}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":2,\"y\":15}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":7,\"y\":15}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":2,\"x\":3,\"y\":16}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":3,\"x\":0,\"y\":10}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":17}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":4,\"y\":17}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":19}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":2,\"x\":7,\"y\":18}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadConsidering\":{\"play_index\":0}}]}";


struct Foo {
  counter: usize
}

impl Foo {
  pub fn add(&mut self) {self.counter+=1;}
  pub fn get(&self) -> usize {self.counter}
}

trait FooTrait {
  fn t_add(&mut self);
  fn t_get(&self) -> usize;
}

impl FooTrait for Foo {
  fn t_add(&mut self) {self.add();}
  fn t_get(&self) -> usize {self.get()}
}

pub fn bench_field(c: &mut Criterion) {
  let cache = ShapeCache::new();
  let rules = GameRules::default();
  let p = Arc::new(cache);
  let p2 = p.clone();
  let p3 = p.clone();
  let log: GameLog = serde_json::from_str(BLA).unwrap();
  c.bench_function("deterministic_play_until_over", move |b| b.iter(|| {
    let mut f = GameField::new_with_cache(2, p.clone(), rules);
    while !f.game_over {
      f.place_block_using_play(0);
    }
  }));
  c.bench_function("random_play_until_over", move |b| b.iter(|| {
    let mut  f = GameField::new_with_cache(2, p2.clone(), rules);
    while !f.game_over {
      let p = f.possible_plays.len() as u8;
      f.place_block_using_play(rand::thread_rng().gen_range(0,p));
    }
  }));
  c.bench_function("replay_log", move |b| b.iter(|| {
    let mut f = GameField::new_with_cache(2, p3.clone(), rules);
    f.restore_from_log(&log, false);
  }));
  c.bench_function("traitcost_direct", move |b| b.iter(|| {
    let mut f = Foo {counter:0};
    for _ in 0..99999999 {
      while f.get() < 9999999999 {
        f.add();
      }
    }
  }));
  c.bench_function("traitcost_struct", move |b| b.iter(|| {
    let mut f: Box<dyn FooTrait> = Box::new(Foo {counter:0});
    for _ in 0..99999999 {
      while f.t_get() < 9999999999 {
        f.t_add();
      }
    }
  }));
}