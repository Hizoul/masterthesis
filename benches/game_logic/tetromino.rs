extern crate criterion;

use criterion::Criterion;
use rustyblocks::game_logic::tetromino::{Position,get_block_shape,get_translated_shape};
use rand::prelude::*;

pub fn bench_base_shape(c: &mut Criterion) {
  c.bench_function("block_shape", |b| b.iter(||
    get_block_shape(
      rand::thread_rng().gen_range(0,5),
      rand::thread_rng().gen_range(0,4)
    )));
}
pub fn bench_translated_shape(c: &mut Criterion) {
  c.bench_function("translate_shape", |b| b.iter(|| {
    let mut rng = rand::thread_rng();
    get_translated_shape(
      rng.gen_range(0,5),
      rng.gen_range(0,4),
      &Position::new(rng.gen_range(0,10),rng.gen_range(0,20))
    )
  }));
}