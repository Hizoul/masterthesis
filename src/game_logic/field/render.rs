use super::GameField;
use crate::game_logic::val::{GAMEFIELD_MAX_WIDTH,GAMEFIELD_MAX_WIDTH_USIZE,GAMEFIELD_MAX_HEIGHT_USIZE,GAMEFIELD_MAX_HEIGHT,AMOUNT_OF_UNIQUE_PIECES_USIZE};

pub trait Renderable {
  fn to_field_string(&self) -> String;
  fn to_field_string_with_player(&self, include_current_player: bool) -> String;
  fn to_field_string_with_player_as_u8(&self) -> Vec<u8>;
  fn to_playable_string(&mut self) -> String;
  fn to_reshapeable_array(&mut self, second_player_as_first: bool) -> Vec<i8>;
  fn to_reshaped_array(&mut self) -> Vec<Vec<i8>>;
  fn to_image(&mut self, second_player_as_first: bool) -> Vec<f32>;
  fn converted(&mut self) -> Vec<f32>;
}
pub const MAX: usize = GAMEFIELD_MAX_HEIGHT_USIZE*GAMEFIELD_MAX_WIDTH_USIZE;
pub const MAX_FOR_SCORE: usize = (GAMEFIELD_MAX_HEIGHT_USIZE+1)*GAMEFIELD_MAX_WIDTH_USIZE;
pub const MAX_FOR_BLOCKS: usize = (GAMEFIELD_MAX_HEIGHT_USIZE+3)*GAMEFIELD_MAX_WIDTH_USIZE;
// 1 for scores, 2 and 3 for block amounts of players, 19 more for valid play indicators
pub const AMOUNT_OF_ADDITIONAL_LINES: usize = 22;

impl Renderable for GameField {
  fn to_field_string(&self) -> String {
    self.to_field_string_with_player(false)
  }
  fn to_field_string_with_player(&self, include_current_player: bool) -> String {
    let mut field = String::with_capacity((GAMEFIELD_MAX_WIDTH_USIZE * GAMEFIELD_MAX_HEIGHT_USIZE) +
        GAMEFIELD_MAX_WIDTH_USIZE + 1);
    field.push('\n');
    for ys in 0..GAMEFIELD_MAX_HEIGHT {
      let y = (GAMEFIELD_MAX_HEIGHT - 1) - ys;
      for x in 0..GAMEFIELD_MAX_WIDTH {
        let mut player = "0".to_string();
        'search: for p in self.tetrominos.iter() {
          let mut i = 0;
          for part in self.shape_cache.get_block_shape(p).iter() {
            if part.x == x && part.y == y {
              let to_set = if i == 0 {((p.from + 1) * 2) - 1} else {((p.from + 1) * 2)};
              player = to_set.to_string();
              break 'search;
            }
            i += 1;
          }
        }
        field.push_str(player.as_str());
      }
      field.push('\n');
    }
    if include_current_player {
      field.push_str(self.current_player.to_string().as_str());
    }
    field
  }
  fn to_field_string_with_player_as_u8(&self) -> Vec<u8> {
    let mut field: [u8; 201] = [0; 201];
    for p in self.tetrominos.iter() {
      let mut i = 0;
      for part in self.shape_cache.get_block_shape(p).iter() {
        let to_set = if i == 0 {i+=1;((p.from + 1) * 2) - 1} else {((p.from + 1) * 2)};
        field[((((GAMEFIELD_MAX_HEIGHT_USIZE-1) - part.y as usize) * GAMEFIELD_MAX_WIDTH_USIZE) + part.x as usize)] = to_set as u8;
      }
    }
    field[200] = self.current_player as u8;
    field.to_vec()
  }
  fn to_reshapeable_array(&mut self, second_player_as_first: bool) -> Vec<i8> {
    // first additional line = scores
    // second and third additional lines = blocks per players
    // 190 additional entries for possible combinations
    let mut field_nums: Vec<i8> = Vec::with_capacity((GAMEFIELD_MAX_HEIGHT_USIZE+AMOUNT_OF_ADDITIONAL_LINES) * GAMEFIELD_MAX_WIDTH_USIZE);
    for index in 0..(GAMEFIELD_MAX_HEIGHT_USIZE+AMOUNT_OF_ADDITIONAL_LINES)*GAMEFIELD_MAX_WIDTH_USIZE {
      if index >= MAX {
        if index < MAX_FOR_SCORE {
          let mut corrected_index = index % MAX;
          if corrected_index < (self.players as usize) {
            if second_player_as_first {
              corrected_index = if corrected_index == 0 {1} else {0};
            }
            field_nums.push(self.scores[corrected_index] as i8);
          } else {
            field_nums.push(-1);
          }
        } else if index < MAX_FOR_BLOCKS {
          let mut player = (index % MAX_FOR_SCORE) / AMOUNT_OF_UNIQUE_PIECES_USIZE;
          if player >= (self.players as usize) {
            field_nums.push(-1);
          } else {
            let block = (index % MAX_FOR_SCORE) % AMOUNT_OF_UNIQUE_PIECES_USIZE;
            if second_player_as_first {
              player = if player == 0 {1} else {0};
            }
            field_nums.push(self.piece_counter[player][block]);
          }
        } else {
          field_nums.push(-1);
        }
      } else {
        field_nums.push(0);
      }
    }
    for play in self.possible_plays.iter() {
      field_nums[(MAX_FOR_BLOCKS)+play.global_index as usize] = 1;
    }
    for p in self.tetrominos.iter() {
      let mut i = 0;
      for part in self.shape_cache.get_block_shape(p).iter() {
        let mut from = p.from;
        if second_player_as_first {
          from = if from == 0 {1} else {0};
        }
        let to_set = if i == 0 {
          i += 1;
          ((from + 1) * 2) - 1
        } else {
          ((from + 1) * 2)
        };
        field_nums[((((GAMEFIELD_MAX_HEIGHT_USIZE-1) - part.y as usize) * GAMEFIELD_MAX_WIDTH_USIZE) + part.x as usize)] = to_set;
      }
    }
    // let b = self.current_block;
    // let o = self.current_orientation;
    // let x = self.current_x;
    // let y = self.highest_y(b, o, x);
    // if !self.game_over {
    //   let hover_block = self.shape_cache.get_translated(b, o, x, y);
    //   for part in hover_block.iter() {
    //     field_nums[((((GAMEFIELD_MAX_HEIGHT_USIZE-1) - part.y as usize) * GAMEFIELD_MAX_WIDTH_USIZE) + part.x as usize)] = (self.current_player as u8 + 1) * 2;
    //   }
    // }
    field_nums
  }
  fn to_image(&mut self, second_player_as_first: bool) -> Vec<f32> {
    let mut to_return: Vec<f32> = vec![0.0; 3072];
    let orig = self.to_reshapeable_array(second_player_as_first);
    for i in 0..200 {
      let y = (i as f32 / 10.0).floor() as usize;
      let x = i % 10;
      let mut current = 0;
      let mut val = 0.0;
      if orig[i] == 1 || orig[i] == 2 {
        current = 0;
        val = if orig[i] == 1 {128.0} else {255.0};
      } else if orig[i] == 3 || orig[i] == 4 {
        current = 2;
        val = if orig[i] == 3 {128.0} else {255.0};
      }
      to_return[((y * 32*3) + x*3) + current] = val;
    }
    for i in 200..210 {
      let y = ((i - 200) as f32 / 10.0).floor() as usize;
      let x = (i % 10) + 10;
      to_return[(y*32*3)+x*3+1] = orig[i] as f32 + 124.0;
    }
    for i in 210..230 {
      let y = ((i - 200) as f32 / 10.0).floor() as usize;
      let x = (i % 10) + 10;
      let val = if orig[i] <= 0 {0.0} else {orig[i] as f32 * 50.0};
      to_return[(y*32*3)+x*3+1] = val;
    }
    for i in 230..420 {
      let y = ((i - 200) as f32 / 10.0).floor() as usize;
      let x = (i % 10) + 10;
      let val = if orig[i] == -1 {
         0.0
      } else {
        255.0
      };
      to_return[(y*32*3)+x*3+1] = val;
    }
    to_return
  }
  fn converted(&mut self) -> Vec<f32> {
    let not_yet_reshaped = self.to_reshapeable_array(false);
    let mut converted: Vec<f32> = Vec::new();
    for num in not_yet_reshaped {
      converted.push(num as f32);
    }
    converted
  }
  fn to_reshaped_array(&mut self) -> Vec<Vec<i8>> {
    let not_yet_reshaped = self.to_reshapeable_array(false);
    let mut reshaped = Vec::new();
    let mut current_sub = Vec::new();
    let i = 0;
    for num in not_yet_reshaped {
      current_sub.push(num);
      if i > 0 && i % 10 == 0 {
        reshaped.push(current_sub.clone());
        current_sub.clear();
      }
    }
    reshaped
  }
  fn to_playable_string(&mut self) -> String {
    let mut field_nums: Vec<Vec<i8>> = Vec::with_capacity(GAMEFIELD_MAX_HEIGHT_USIZE);
    for _ in 0..GAMEFIELD_MAX_HEIGHT {
      let mut new_line = Vec::with_capacity(GAMEFIELD_MAX_WIDTH_USIZE);
      for _ in 0..GAMEFIELD_MAX_WIDTH {
        new_line.push(0);
      }
      field_nums.push(new_line);
    }
    for p in self.tetrominos.iter() {
      for part in self.shape_cache.get_block_shape(p).iter() {
        field_nums[(GAMEFIELD_MAX_HEIGHT_USIZE-1) - part.y as usize][part.x as usize] = ((p.from + 1) * 2) - 1;
      }
    }
    let b = self.current_block;
    let o = self.current_orientation;
    let x = self.current_x;
    let y = self.highest_y(b, o, x);
    if !self.game_over {
      let hover_block = self.shape_cache.get_translated(b, o, x, y);
      for part in hover_block.iter() {
        let new_y = (GAMEFIELD_MAX_HEIGHT_USIZE-1) - part.y as usize;
        let new_x = part.x as usize;
        if new_x < GAMEFIELD_MAX_WIDTH_USIZE && new_y < GAMEFIELD_MAX_HEIGHT_USIZE {
          field_nums[new_y][new_x] = (self.current_player + 1) * 2;
        }
      }
    }
    let mut field = String::with_capacity((GAMEFIELD_MAX_WIDTH_USIZE * GAMEFIELD_MAX_HEIGHT_USIZE) +
        GAMEFIELD_MAX_WIDTH_USIZE + 1);
    for top in field_nums.iter() {
      for ele in top.iter() {
        field.push_str(ele.to_string().as_str());
      }
      field.push('\n');
    }
    field
  }
}