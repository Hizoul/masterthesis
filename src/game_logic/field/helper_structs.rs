use serde::{Deserialize, Serialize};
use crate::game_logic::tetromino::{Tetromino};
use std::sync::{Arc};

#[derive(Debug,Clone)]
pub struct Touching {
  pub block: Arc<Tetromino>,
  pub touching: Vec<Arc<Tetromino>>,
  pub connected: Vec<Arc<Tetromino>>
}

impl PartialEq for Touching {
  fn eq(&self, other: &Touching) -> bool {
    self.block == other.block
  }
}

impl Touching {
  pub fn new(block: Arc<Tetromino>) -> Touching {
    return Touching {
      block, touching: Vec::with_capacity(4), connected: Vec::with_capacity(4)
    }
  }
}

#[derive(Debug,Serialize,Deserialize,Clone)]
pub struct OwnedPosition {
    pub x: i8,
    pub y: i8,
    pub from: i8
}

#[derive(Debug,Serialize,Deserialize,Copy,Clone)]
pub struct GameRules {
  pub roll_dice: bool,
  pub limit_minus: bool
}

impl GameRules {
  pub fn original() -> GameRules {
    GameRules {roll_dice:true,limit_minus:true}
  }
  pub fn original_without_limit() -> GameRules {
    GameRules {roll_dice:true,limit_minus:false}
  }
  pub fn deterministic() -> GameRules {
    GameRules {roll_dice:false,limit_minus:true}
  }
  pub fn deterministic_without_limit() -> GameRules {
    GameRules {roll_dice:false,limit_minus:false}
  }
}

impl Default for GameRules {
  fn default() -> GameRules {
    GameRules::original_without_limit()
  }
}