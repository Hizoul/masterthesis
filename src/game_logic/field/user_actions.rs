use super::GameField;
use crate::game_logic::tetromino::{get_rotation_amount};
use crate::game_logic::log::{Play};
use crate::game_logic::val::{GAMEFIELD_MAX_WIDTH,AMOUNT_OF_UNIQUE_PIECES};

pub const ACTION_LEFT: u8 = 0;
pub const ACTION_RIGHT: u8 = 1;
pub const ACTION_ROTATE_LEFT: u8 = 2;
pub const ACTION_ROTATE_RIGHT: u8 = 3;
pub const ACTION_PREV_BLOCK: u8 = 4;
pub const ACTION_NEXT_BLOCK: u8 = 5;
pub const ACTION_PLACE_BLOCK: u8 = 6;

pub trait UserActions {
  fn do_user_action(&mut self, user_action: u8);
  fn find_action(&self, block: i8, orientation: i8, x: i8) -> Option<&Play>;
}

impl UserActions for GameField {
  fn find_action(&self, block: i8, orientation: i8, x: i8) -> Option<&Play> {
    let found_action = self.possible_plays.iter().find(|play| {
      play.x == x && play.block == block && play.rotation == orientation
    });
    found_action
  }
  fn do_user_action(&mut self, user_action: u8) {
    match user_action {
      ACTION_PLACE_BLOCK => {
        let action: Option<&Play> = self.find_action(self.current_block,self.current_orientation,self.current_x);
        if action.is_some() {
          let relevant_index = action.unwrap().index;
          self.place_block_using_play(relevant_index);
        }
      }
      ACTION_LEFT => {
        self.current_x = self.current_x - 1;
        if self.current_x < 0 {
          self.current_x = GAMEFIELD_MAX_WIDTH - 1;
        }
      }
      ACTION_RIGHT => {
        self.current_x = self.current_x + 1;
        if self.current_x >= GAMEFIELD_MAX_WIDTH {
          self.current_x = 0;
        }
      }
      ACTION_ROTATE_LEFT => {
        self.current_orientation = self.current_orientation - 1;
        if self.current_orientation < 0 {
          self.current_orientation = get_rotation_amount(self.current_block) - 1; 
        }
      }
      ACTION_ROTATE_RIGHT => {
        self.current_orientation = self.current_orientation + 1;
        if self.current_orientation >= get_rotation_amount(self.current_block) {
          self.current_orientation = 0; 
        }
      }
      ACTION_PREV_BLOCK => {
        self.current_block = self.current_block + 1;
        if self.current_block >= AMOUNT_OF_UNIQUE_PIECES {
          self.current_block = 0; 
        }
      }
      ACTION_NEXT_BLOCK => {
        self.current_block = self.current_block + 1;
        if self.current_block >= AMOUNT_OF_UNIQUE_PIECES {
          self.current_block = 0; 
        }
      }
      _ => {
      }
    }
  }
}