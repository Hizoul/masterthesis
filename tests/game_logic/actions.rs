use rustyblocks::game_logic::field::{GameField};
use rustyblocks::game_logic::field::render::Renderable;
use rustyblocks::game_logic::field::helper_structs::{GameRules};
use rustyblocks::game_logic::field::user_actions::UserActions;
use rustyblocks::game_logic::field::user_actions::{ACTION_LEFT,ACTION_RIGHT,ACTION_ROTATE_LEFT,ACTION_ROTATE_RIGHT,ACTION_NEXT_BLOCK,ACTION_PLACE_BLOCK};
use insta::{assert_snapshot};

#[test]
fn game_field_actions() {
  let mut g = GameField::new_with_rules(2, GameRules::deterministic());
  assert_snapshot!("field_before_anything", g.to_playable_string());
  g.do_user_action(ACTION_LEFT);
  assert_snapshot!("field_after_move_left_over_bounds", g.to_playable_string());
  g.do_user_action(ACTION_RIGHT);
  assert_snapshot!("field_after_move_right_over_bounds", g.to_playable_string());
  g.do_user_action(ACTION_ROTATE_LEFT);
  assert_snapshot!("field_after_rotate_left_over_bounds", g.to_playable_string());
  g.do_user_action(ACTION_ROTATE_RIGHT);
  assert_snapshot!("field_after_rotate_right_over_bounds", g.to_playable_string());
  g.do_user_action(ACTION_PLACE_BLOCK);
  assert_snapshot!("field_after_place_block", g.to_playable_string());
  g.do_user_action(ACTION_RIGHT);
  g.do_user_action(ACTION_NEXT_BLOCK);
  assert_snapshot!("field_after_switch_block_to_square", g.to_playable_string());
  g.do_user_action(ACTION_NEXT_BLOCK);
  assert_snapshot!("field_after_switch_block_to_t", g.to_playable_string());
  g.do_user_action(ACTION_NEXT_BLOCK);
  assert_snapshot!("field_after_switch_block_to_s", g.to_playable_string());
  g.do_user_action(ACTION_NEXT_BLOCK);
  assert_snapshot!("field_after_switch_block_to_l", g.to_playable_string());
  g.do_user_action(ACTION_NEXT_BLOCK);
  assert_snapshot!("field_after_switch_block_back_to_i", g.to_playable_string());
}