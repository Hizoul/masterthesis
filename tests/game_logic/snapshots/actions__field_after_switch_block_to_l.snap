---
created: "2019-09-18T11:15:47.019231149Z"
creator: insta@0.8.1
source: tests/game_logic/actions.rs
expression: g.to_playable_string()
---
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0000000000
0004000000
0444000000
1111000000

